chainimport/interfaces.go

```diff
@@ -20,6 +23,53 @@ type Header interface {
 	Deserialize(io.Reader, uint32) error
 }
 
+// StatelessIterator provides functional iteration over headers using Go's iter
+// package. Each method call creates a fresh, independent iterator.
+type StatelessIterator interface {
+	// Iterator returns an iter.Seq2[header, error] for the specified range.
+	// Each call creates a fresh iterator that independently traverses the
+	// range.
+	Iterator(start, end uint32) iter.Seq2[Header, error]
+
+	// BatchIterator returns an iterator that yields batches of headers for
+	// the specified range, where each batch respects the configured batch
+	// size limit. Each call creates a fresh iterator that independently
+	// traverses the range.
+	BatchIterator(start, end uint32) iter.Seq2[[]Header, error]
+
+	// ReadBatch collects all headers from the given range into a slice.
+	ReadBatch(start, end uint32) ([]Header, error)
+
+	// Close releases any resources associated with the iterator.
+	Close() error
+}
+
+// StatefulIterator provides position-based iteration over headers with internal
+// state.
+type StatefulIterator interface {
+	// Next returns the next header in the current iteration sequence.
+	// Maintains internal position state. Returns nil, io.EOF when no more
+	// headers are available.
+	Next() (Header, error)
```

Comment:
Any reason this doesn't just continue to use `iter.Seq` like the above? 

From my PoV, the main difference between a stateful and stateless iterator is that a stateless one doesn't maintain any closure state, while a stateful one would. As is this misses out on the return value returned from normal iterators that signal when to stop iteration. 

`iter.Pull` can also be used to turn a normal iterator into one that can be incrementally advanced. 

Still getting through the PR, so just an initial thought.